## Table of Contents

- Series
    - Creating a Series
    - Sorting a Series
        - 1.0 - Now Try This
    - Accessing an element in a Series
        - 2.0 - Now Try This
        - 3.0 - Now Try This
    - Binary Operations
        - 4.0 - Now Try This

- Advanced Dataframe
    - Creaing new columns
    - Sorting the dataset
    - Selecting subset of data
        - 5.0 - Now Try This
    - Dropping entities
    - Unique values in columns
    - Joining dataset
        - Merge
        - Join

- Tutorial
    - Join dataset
    - Groupby
        - 1.0 - Now Try This
        - agg()
    - Filtering based on condition
        - 2.0 - Now Try This
        - Pop-up Question


## Dataframe vs. Series
There are two main data structures you need to get comfortable with: **Dataframe** and **Series** and we have covered the basics of DataFrame in Week 3. 

To refresh your memory, **DataFrame** is a tabular, column-oriented data structure with both row and column labels. 

**Series** is 1-dimensional labelled array and it can hold data of any type (integer, string, float, python objects, etc.). Its labels are called an index. You can think of a Series as a ordered dictionary, as it is a mapping of index values to data values.


## Series

- Creating a Series
- Sorting a Series
- Accessing an element in a Series
- Binary Operations (add, sum, mul, etc.)


### Creating a Series

A series can be created from 
1. scalar
2. list (array) 
2. dictionary

First and foremost, let's import Pandas library.

In [36]:
import pandas as pd

#### 1. Series from a scalar value

In [37]:
scalar_series = pd.Series(5, index =[0, 1, 2, 3, 4, 5]) 

#### 2-1. Series from a list. 

By default, index starts at 0 and ends at len(array)-1. 

In this example, as we didn't specify the index, the first index is 0 and the last index is 4, which is len(array)-1.

In [38]:
import numpy as np
array = np.array(['a','b','c','d'])
pd.Series(array)

0    a
1    b
2    c
3    d
dtype: object

#### 2-2. Series from a list with index

If we want to assign a different index instead of the default ones, we can specify the index as well.

In [39]:
array = np.array(['a','b','c','d'])
pd.Series(array,index=[1,2,3,4])

1    a
2    b
3    c
4    d
dtype: object

We can change the order of the elements in the array. (d, b, c, a)
But, the Series is sorted in the specified order of index!

In [40]:
array = np.array(['d','b','c','a'])
pd.Series(array,index=[1,2,3,4])

1    d
2    b
3    c
4    a
dtype: object

Using the examples below, we can note that Series is NOT ordered by value or index.

Instead, Series is ordered in the specified order of indices: 105 --> 103 --> 102 --> 109 --> 104 --> 110

In [41]:
names = np.array(['Daisy', 'Matt', 'Kelly', 'Mike', 'Ashley', 'Kyle'])
pd.Series(names, index=[105,103,102,109,104,110])

105     Daisy
103      Matt
102     Kelly
109      Mike
104    Ashley
110      Kyle
dtype: object

Then, the question is "Are the indices unique?"
Let's check it out!

In [42]:
names = np.array(['Daisy', 'Matt', 'Kelly', 'Mike', 'Ashley', 'Kyle'])
pd.Series(names, index=[100,100,102,102,103,110])

100     Daisy
100      Matt
102     Kelly
102      Mike
103    Ashley
110      Kyle
dtype: object

The indices don't have to be unique!

Can we use non-numerical indices?
Let's check it out!

In [43]:
names = np.array(['Daisy', 'Matt', 'Kelly', 'Mike'])
cities=["Atlanta", "San Francisco", "New York", "Seattle"]

pd.Series(names, index=cities)

Atlanta          Daisy
San Francisco     Matt
New York         Kelly
Seattle           Mike
dtype: object

We found out that the indices don't have to numbers; indices could be of any types! 

In [44]:
names = np.array(['Daisy', 'Matt', 'Kelly', 'Mike'])
pd.Series(names, index=[0.79, [0.8, 1.4], 13, "Hello"])

0.79          Daisy
[0.8, 1.4]     Matt
13            Kelly
Hello          Mike
dtype: object

We used float, list, integer, and string as indices and it worked!

This example is to demonstrate that you can use any types as indices; in real worlds, it's more common to keep the datatype consistent for indices.

#### 3. Series from a dict
We can create a Series from a **dict**! Let's see how it works!

In [45]:
# aDict stores how many fruits we have.
aDict = {'Apple':3, 'Banana':5, 'Cherry': 2, 'Mango': 13, 'Peach': 10}
pd.Series(aDict)

Apple      3
Banana     5
Cherry     2
Mango     13
Peach     10
dtype: int64

When we are passing a dict, the order in the dict is preserved.

In [46]:
aDict = {'Mango': 13, 'Banana':5, 'Cherry':2, 'Apple': 3, 'Peach': 10}
pd.Series(aDict)

Mango     13
Banana     5
Cherry     2
Apple      3
Peach     10
dtype: int64

If we want to know the datatype of Series, we can call dtype function.

In [47]:
aDict = {'Mango': 13, 'Banana':5, 'Cherry':2, 'Apple': 3, 'Peach': 10}
pd.Series(aDict).dtype

dtype('int64')

Now, let's change the value for "Mango." The value is changed to 13.45 and see what dtype returns!

In [48]:
aDict = {'Mango': 13.45, 'Banana':5, 'Cherry':2, 'Apple': 3, 'Peach': 10}
pd.Series(aDict).dtype

dtype('float64')

If we want to print out indices only, we can call index function.

In [49]:
aDict = {'Mango': 13.45, 'Banana':5, 'Cherry':2, 'Apple': 3, 'Peach': 10}
pd.Series(aDict).index

Index(['Mango', 'Banana', 'Cherry', 'Apple', 'Peach'], dtype='object')

### Sorting a Series

If we want to sort the Series in the order of "key" or "index", we can use sort_index() function.

In [50]:
s = pd.Series(aDict)
s.sort_index()

Apple      3.00
Banana     5.00
Cherry     2.00
Mango     13.45
Peach     10.00
dtype: float64

Similarly, if we want to sort the series based on the value, we can use a built-in function sort_values. By default, it will sort the elements in an ascending order.

In [51]:
s = pd.Series(aDict)
s.sort_values()

Cherry     2.00
Apple      3.00
Banana     5.00
Peach     10.00
Mango     13.45
dtype: float64

Let's sort the series in a descending order.

In [52]:
s.sort_values(ascending=False)

Mango     13.45
Peach     10.00
Banana     5.00
Apple      3.00
Cherry     2.00
dtype: float64

### 1.0 - Now Try This
Sort the series by index in a descending order.

In [53]:
s.sort_index(ascending=False)

Peach     10.00
Mango     13.45
Cherry     2.00
Banana     5.00
Apple      3.00
dtype: float64

### Accessing an element in a Series

We can access elements in a Series in the following ways:
- by index number
- by index label

In [54]:
array = np.array(['a','b','c','d','e','f','g','h'])
s = pd.Series(array)

#### 1. By index number

We can access the first element by index number 0.

In [55]:
s[0]

'a'

In [56]:
s[5]

'f'

We can access the first 3 elements with index operation [:3]

In [57]:
s[:3]

0    a
1    b
2    c
dtype: object

### 2.0 - Now Try This

Access the items in the index of 3-5 (inclusive) using index operation.

In [58]:
s[3:6]

3    d
4    e
5    f
dtype: object

#### By Index Label


We can retrieve a single element using index label.

In [59]:
aDict = {'Apple':3, 'Banana':5, 'Cherry': 2, 'Peach': 10}
s = pd.Series(aDict)
s['Apple']

3

In [60]:
s['Peach']

10

### 3.0 - Now Try This



In [61]:
s = pd.Series([1,2,3,4,5],index = ['a','b','c','d','e'])

Retrieve an element that corresponds to index 'd'

In [62]:
s['d']

4

### Binary Operations of Series
- Add
- Sub
- Mul

Here, we have created two Series with numbers.

In [63]:
# creating a first series
s1 = pd.Series([1,5,6,2])
 
# creating a second series
s2 = pd.Series([4,1,3,5])

Let's add the two Series!

In [64]:
# answer = s1 + s2
answer = s1.add(s2)
print(answer)

0    5
1    6
2    9
3    7
dtype: int64


We can also subtract a Series from another Series using sub() function.

In [65]:
# answer = s1 - s2
answer = s1.sub(s2)
print(answer)

0   -3
1    4
2    3
3   -3
dtype: int64


In [66]:
answer = s1.mul(s2)
print(answer)

0     4
1     5
2    18
3    10
dtype: int64


In a simpler approach, we can use mathematical notation and it will result in the same output.

In [67]:
s1 + s2

0    5
1    6
2    9
3    7
dtype: int64

In [68]:
s1 - s2

0   -3
1    4
2    3
3   -3
dtype: int64

In [69]:
s1 * s2

0     4
1     5
2    18
3    10
dtype: int64

Now, let's create two Series with index. Note that the order of index is different in two Series.

In [70]:
# creating a first series
s1 = pd.Series([1,5,6,2], index=['b','c','d','a'])
 
# creating a second series
s2 = pd.Series([4,1,3,5], index=['a','b','c','d'])

In [71]:
s1 + s2

a     6
b     2
c     8
d    11
dtype: int64

s['a'] = 6 was obtained by s1['a'] + s2['a'] and all the other values are calculated in the same way.

### 4.0 - Now Try This

Write down what this code will return on a piece of paper.
Compare your answer with the output of the code.

In [72]:
s1 * s2

a     8
b     1
c    15
d    30
dtype: int64

## Advanced Dataframe

Let's move onto dataframe and create a dataframe on people's income.

In [73]:
data = {'First Name': ['Jennifer', 'Chris', 'John', 'Annie', 'Chloe', 'Jeremy', 'Allison'], 
        'Last Name': ['Brown', 'Smith', 'Williams', 'Wong', 'Anderson', 'Scott', 'Kim'], 
        'Age': [16, 32, 21, 35, 27, 42, 28],
        'Hourly Wage': [8,14,60,44,80,54,32],
        'Hours per week': [20,28,40,40,32,40,40]
        } 

emp = pd.DataFrame(data)

### Creating new columns

#### String Manipulation

Let's combine First Name and Last Name and create a new column called "Full Name."

The ```' '``` in the middle is adding a space between the first name and last name. 

In [74]:
# string manipulation
emp['Full Name'] = emp['First Name'] + ' ' + emp['Last Name']
emp

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name
0   Jennifer     Brown   16            8              20  Jennifer Brown
1      Chris     Smith   32           14              28     Chris Smith
2       John  Williams   21           60              40   John Williams
3      Annie      Wong   35           44              40      Annie Wong
4      Chloe  Anderson   27           80              32  Chloe Anderson
5     Jeremy     Scott   42           54              40    Jeremy Scott
6    Allison       Kim   28           32              40     Allison Kim

#### Numerical column

Let's calculate weekly salary for each person and add a column called ```Weekly Salary```. We need to do a simple math to get the weekly salary.

In [75]:
# multiplication of two numerical columns
emp['Weekly Salary'] = emp['Hourly Wage'] * emp['Hours per week']
emp

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
0   Jennifer     Brown   16            8              20  Jennifer Brown   
1      Chris     Smith   32           14              28     Chris Smith   
2       John  Williams   21           60              40   John Williams   
3      Annie      Wong   35           44              40      Annie Wong   
4      Chloe  Anderson   27           80              32  Chloe Anderson   
5     Jeremy     Scott   42           54              40    Jeremy Scott   
6    Allison       Kim   28           32              40     Allison Kim   

   Weekly Salary  
0            160  
1            392  
2           2400  
3           1760  
4           2560  
5           2160  
6           1280

Now, we've added two columns "Full Name" and "Weekly Salary."

#### List Comprehension

Let's create a new column called ```High Income``` based on the ```Weekly Salary```. We will use a list comprehension for this (you can refer to Week 2 materials if you can't recall list comprehenstion.)

```High Income``` will be ```True``` if ```Weekly Salary``` is > 2000, otherwise False.

In [76]:
# list comprehension
emp['High Income'] = [True if x > 2000 else False for x in emp['Weekly Salary']]
emp

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
0   Jennifer     Brown   16            8              20  Jennifer Brown   
1      Chris     Smith   32           14              28     Chris Smith   
2       John  Williams   21           60              40   John Williams   
3      Annie      Wong   35           44              40      Annie Wong   
4      Chloe  Anderson   27           80              32  Chloe Anderson   
5     Jeremy     Scott   42           54              40    Jeremy Scott   
6    Allison       Kim   28           32              40     Allison Kim   

   Weekly Salary  High Income  
0            160        False  
1            392        False  
2           2400         True  
3           1760        False  
4           2560         True  
5           2160         True  
6           1280        False

### Sorting the dataset

Let's sort the data by 'Weekly Salary'

In [77]:
emp.sort_values(by='Weekly Salary')

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
0   Jennifer     Brown   16            8              20  Jennifer Brown   
1      Chris     Smith   32           14              28     Chris Smith   
6    Allison       Kim   28           32              40     Allison Kim   
3      Annie      Wong   35           44              40      Annie Wong   
5     Jeremy     Scott   42           54              40    Jeremy Scott   
2       John  Williams   21           60              40   John Williams   
4      Chloe  Anderson   27           80              32  Chloe Anderson   

   Weekly Salary  High Income  
0            160        False  
1            392        False  
6           1280        False  
3           1760        False  
5           2160         True  
2           2400         True  
4           2560         True

Now, all of the high income workers are located in the bottom of the dataset.

### Selecting subset of data based on condition

Let's review how we can select subset of data based on criteria.

Selecting rows where 'High Income' is True

In [78]:
emp[emp['High Income'] == True]

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
2       John  Williams   21           60              40   John Williams   
4      Chloe  Anderson   27           80              32  Chloe Anderson   
5     Jeremy     Scott   42           54              40    Jeremy Scott   

   Weekly Salary  High Income  
2           2400         True  
4           2560         True  
5           2160         True

### 5.0 - Now Try This

Let's find people who are younger than 30 years old AND have high income.

(Hint: Make sure to use ```()``` on each condition.)

In [79]:
emp[(emp['High Income'] == True) & (emp['Age'] < 30)]

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
2       John  Williams   21           60              40   John Williams   
4      Chloe  Anderson   27           80              32  Chloe Anderson   

   Weekly Salary  High Income  
2           2400         True  
4           2560         True

### Dropping rows

There are two ways of dropping rows.

1. ```df.drop(index #)```
2. ```df.drop(df[<some boolean condition>].index)```: 

The first code is used when we know exactly what rows we want to drop from the dataframe.
With the second code, we can drop rows that do not meet certain criteria.

Let's drop rows where 'High Income' is False.

In [80]:
emp.drop(emp[emp['High Income'] == False].index)

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
2       John  Williams   21           60              40   John Williams   
4      Chloe  Anderson   27           80              32  Chloe Anderson   
5     Jeremy     Scott   42           54              40    Jeremy Scott   

   Weekly Salary  High Income  
2           2400         True  
4           2560         True  
5           2160         True

In [81]:
emp

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
0   Jennifer     Brown   16            8              20  Jennifer Brown   
1      Chris     Smith   32           14              28     Chris Smith   
2       John  Williams   21           60              40   John Williams   
3      Annie      Wong   35           44              40      Annie Wong   
4      Chloe  Anderson   27           80              32  Chloe Anderson   
5     Jeremy     Scott   42           54              40    Jeremy Scott   
6    Allison       Kim   28           32              40     Allison Kim   

   Weekly Salary  High Income  
0            160        False  
1            392        False  
2           2400         True  
3           1760        False  
4           2560         True  
5           2160         True  
6           1280        False

Even though we dropped the columns where High Income == False, emp remains unchanged.

This is because we didn't update emp!

We can update the dataframe once we drop the rows by setting ```inplace=True```

In [82]:
emp.drop(emp[emp['High Income'] == False].index, inplace=True)

In [83]:
emp

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
2       John  Williams   21           60              40   John Williams   
4      Chloe  Anderson   27           80              32  Chloe Anderson   
5     Jeremy     Scott   42           54              40    Jeremy Scott   

   Weekly Salary  High Income  
2           2400         True  
4           2560         True  
5           2160         True

Here we go! emp has been updated!

### Company Dataframe

Here, we have the dataset with the company name each person in ```emp``` works for.

In [84]:
data = {'First Name': ['Jennifer', 'Chris', 'John', 'Annie', 'Chloe', 'Jeremy', 'Allison'], 
        'Last Name': ['Brown', 'Smith', 'Williams', 'Wong', 'Anderson', 'Scott', 'Kim'], 
        'Company': ['Home Depot', 'Bit Project', 'Microsoft', 'Bit Project', 'Disney', 'Adidas', 'Home Depot']
        } 

companies = pd.DataFrame(data)

In [85]:
companies

First Name Last Name      Company
0   Jennifer     Brown   Home Depot
1      Chris     Smith  Bit Project
2       John  Williams    Microsoft
3      Annie      Wong  Bit Project
4      Chloe  Anderson       Disney
5     Jeremy     Scott       Adidas
6    Allison       Kim   Home Depot

### Unique values in columns

Let's find out the unique company names in company dataframe.

```df[[column_name]].drop_duplicates()```

This code will literally drop duplicates in the ```column_name```!

In [86]:
companies[["Company"]].drop_duplicates()

Company
0   Home Depot
1  Bit Project
2    Microsoft
4       Disney
5       Adidas

To find out the number of unique values in a column, we can use ```nunique()```.

```nunique()``` can be interpreted as **n**umber of **unique** values.

In [87]:
companies["Company"].nunique()

5

These two functions will be very helpful when working with a large, complicated dataset!

### Joining Dataset

Let's join the two tables ```emp``` and ```companies``` to find out which companies those people work for.

### 1. Merge

These are the parameters we can use when we merge two dataframes.

```DataFrame.merge(right, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, sort=False, suffixes='_x', '_y', copy=True, indicator=False, validate=None)```

- how: {‘inner’, ‘left’, ‘right’, ‘outer’}, default ‘inner’
- on: Column or index level names to join on. These must be found in both DataFrames.
- left_on: Column or index level names to join on in the left DataFrame.
- right_on: Column or index level names to join on in the right DataFrame.
- left_index: Use the index from the left DataFrame as the join key.
- right_index: Use the index from the right DataFrame as the join key
- suffixes: A length-2 sequence where each element is optionally a string indicating the suffix to add to overlapping column names in left and right respectively. 

We will cover the most fundamental and commonly used ones here. Feel free to check for more details at https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html 


#### Merge - How

```how``` indicates a type of ```join``` and there are 4: ```inner```, ```left```, ```right```, and ```outer```.

![inner_join.png](assets/inner_join.png)

```inner join``` will take in the common items / columns between the two dataframes.

![left_join.png](assets/left_join.png)

```left join``` will include everything from left dataframe and add a column from right dataframe where there's a match in the right dataframe.

![right_join.png](assets/right_join.png)

Similar to ```left join```, ```right join``` will include everything from right dataframe and add a column from left dataframe where there's a match in the left dataframe.

![full_join.png](assets/full_join.png)

```full join``` will include everything from left and right dataframe even if there's no common item between the two dataframes.

Understanding different types of ```join```s are very important as this will help us to determine which type of ```join``` we will perform on a dataframe.

#### Merge - How Example:

We can merge ```emp``` that has weekly salary information and ```companies``` that has the name of the company each person works for.

The code below will return a dataframe of two merged objects.

In [89]:
pd.merge(emp, companies)

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
0       John  Williams   21           60              40   John Williams   
1      Chloe  Anderson   27           80              32  Chloe Anderson   
2     Jeremy     Scott   42           54              40    Jeremy Scott   

   Weekly Salary  High Income    Company  
0           2400         True  Microsoft  
1           2560         True     Disney  
2           2160         True     Adidas

Yay, it worked! 

We dropped the rows from ```emp``` where employee's weekly salary is less than $2000, so the updated ```emp``` has 3 entities.

By default, ```merge``` performs ```inner join``` and that's why it's showing less number of rows than ```companies``` dataframe.

Now, let's try ```right_join``` and see how it works!

In [92]:
pd.merge(emp, companies, how='right')

First Name Last Name   Age  Hourly Wage  Hours per week       Full Name  \
0       John  Williams  21.0         60.0            40.0   John Williams   
1      Chloe  Anderson  27.0         80.0            32.0  Chloe Anderson   
2     Jeremy     Scott  42.0         54.0            40.0    Jeremy Scott   
3   Jennifer     Brown   NaN          NaN             NaN             NaN   
4      Chris     Smith   NaN          NaN             NaN             NaN   
5      Annie      Wong   NaN          NaN             NaN             NaN   
6    Allison       Kim   NaN          NaN             NaN             NaN   

   Weekly Salary High Income      Company  
0         2400.0        True    Microsoft  
1         2560.0        True       Disney  
2         2160.0        True       Adidas  
3            NaN         NaN   Home Depot  
4            NaN         NaN  Bit Project  
5            NaN         NaN  Bit Project  
6            NaN         NaN   Home Depot

As our right dataframe is ```companies``` with 7 entities, it shows all of the 7 employees in ```companies``` dataframe.

But, as we've dropped the rows that do not meet our salary expectation in ```emp```, the result of ```right_join``` is missing ```Age```, ```Hourly Wage```, ```Hours per week```, etc for some employees.

#### Merge - On, left_on, right_on

```on```, ```left_on```, and ```right_on``` indicate a column or index level names to join on.

- ```on``` is used when the same column name exists in both dataframes.
- ```left_on``` and ```right_on``` is used when the column names are different in two dataframes.

#### Merge - On Example

Let's join the two dataframes using common column names, ```First Name``` and ```Last Name```.

In [93]:
pd.merge(emp, companies, on=['First Name', 'Last Name'])

First Name Last Name  Age  Hourly Wage  Hours per week       Full Name  \
0       John  Williams   21           60              40   John Williams   
1      Chloe  Anderson   27           80              32  Chloe Anderson   
2     Jeremy     Scott   42           54              40    Jeremy Scott   

   Weekly Salary  High Income    Company  
0           2400         True  Microsoft  
1           2560         True     Disney  
2           2160         True     Adidas

The resulting dataset shows three entities, since ```inner join``` was performed by default.

#### Merge - left_on, right_on Example

It might happen that the two dataframes we want to merge have different column names.
In this case, we will have to specify the left column name and the right column name.

Let's join the two dataframes using a different column name. 

- ```emp``` column name: ```Full Name``` 
- ```companies``` column name: ```Name```

In [94]:
companies['Name'] = companies['First Name'] + ' ' + companies['Last Name']

In [95]:
pd.merge(emp, companies, left_on='Full Name', right_on='Name')

First Name_x Last Name_x  Age  Hourly Wage  Hours per week       Full Name  \
0         John    Williams   21           60              40   John Williams   
1        Chloe    Anderson   27           80              32  Chloe Anderson   
2       Jeremy       Scott   42           54              40    Jeremy Scott   

   Weekly Salary  High Income First Name_y Last Name_y    Company  \
0           2400         True         John    Williams  Microsoft   
1           2560         True        Chloe    Anderson     Disney   
2           2160         True       Jeremy       Scott     Adidas   

             Name  
0   John Williams  
1  Chloe Anderson  
2    Jeremy Scott

As mentioned earlier, by default, ```inner join``` has been applied and the resulting dataframe show 3 entities!

In the result dataset, ```First Name_x``` is from the first table, which is ```emp``` while ```First Name_y``` is from the second table, which is ```companies```.

#### Merge - Suffixes

Suffixes indicate the suffix to add to overlapping column names in left and right respectively.

In the previous example, by default, suffixes ```_x``` and ```_y``` were added respectively.

We can specify suffixes by setting ```suffixes=[]```.

#### Merge - Suffixes Example

Let's add ```_1``` to the first dataframe and ```_2``` to the second dataframe.

In [96]:
pd.merge(emp, companies, left_on='Full Name', right_on='Name', suffixes=['_1', '_2'])

First Name_1 Last Name_1  Age  Hourly Wage  Hours per week       Full Name  \
0         John    Williams   21           60              40   John Williams   
1        Chloe    Anderson   27           80              32  Chloe Anderson   
2       Jeremy       Scott   42           54              40    Jeremy Scott   

   Weekly Salary  High Income First Name_2 Last Name_2    Company  \
0           2400         True         John    Williams  Microsoft   
1           2560         True        Chloe    Anderson     Disney   
2           2160         True       Jeremy       Scott     Adidas   

             Name  
0   John Williams  
1  Chloe Anderson  
2    Jeremy Scott

Since inner join was performed, only 3 common entities show up and there's actually no need to indicate ```_1``` or ```_2```.

Let's try right join and see how dataframe looks different!

In [97]:
pd.merge(emp, companies, how = 'right', left_on='Full Name', right_on='Name', suffixes=['_emp', '_companies'])

First Name_emp Last Name_emp   Age  Hourly Wage  Hours per week  \
0           John      Williams  21.0         60.0            40.0   
1          Chloe      Anderson  27.0         80.0            32.0   
2         Jeremy         Scott  42.0         54.0            40.0   
3            NaN           NaN   NaN          NaN             NaN   
4            NaN           NaN   NaN          NaN             NaN   
5            NaN           NaN   NaN          NaN             NaN   
6            NaN           NaN   NaN          NaN             NaN   

        Full Name  Weekly Salary High Income First Name_companies  \
0   John Williams         2400.0        True                 John   
1  Chloe Anderson         2560.0        True                Chloe   
2    Jeremy Scott         2160.0        True               Jeremy   
3             NaN            NaN         NaN             Jennifer   
4             NaN            NaN         NaN                Chris   
5             NaN            NaN         NaN                Annie   
6             NaN            NaN         NaN              Allison   

  Last Name_companies      Company            Name  
0            Williams    Microsoft   John Williams  
1            Anderson       Disney  Chloe Anderson  
2               Scott       Adidas    Jeremy Scott  
3               Brown   Home Depot  Jennifer Brown  
4               Smith  Bit Project     Chris Smith  
5                Wong  Bit Project      Annie Wong  
6                 Kim   Home Depot     Allison Kim

Because we specified right join, it shows all of the entities from ```companies``` dataframe and matching entities from ```emp``` dataframe.

The ```First Name_emp``` and ```Last Name_emp``` show three entities only while ```First Name_companies``` and ```Last Name_companies``` show all entities!

### 2. Join

These are the parameters we can use when we merge two dataframes.

```DataFrame.join(other, on=None, how='left', lsuffix='', rsuffix='', sort=False)```

- other: DataFrame, Series, or list of DataFrame to join
- on: Column or index level names to join on the index in other. These must be found in both DataFrames.
- how: {‘inner’, ‘left’, ‘right’, ‘outer’}, default ‘inner’
- lsuffix: Suffix to use from left frame’s overlapping columns.
- rsuffix: Suffix to use from right frame’s overlapping columns.

We have covered ```on``` and ```how``` in detail in ```merge```.

So, we will try one example of ```join``` using ```lsuffix``` and ```rsuffix```.

#### Join - lsuffix and rsuffix

This is similar to ```suffixes``` in ```merge```.

In [98]:
emp.join(companies, lsuffix='_1', rsuffix='_2')

First Name_1 Last Name_1  Age  Hourly Wage  Hours per week       Full Name  \
2         John    Williams   21           60              40   John Williams   
4        Chloe    Anderson   27           80              32  Chloe Anderson   
5       Jeremy       Scott   42           54              40    Jeremy Scott   

   Weekly Salary  High Income First Name_2 Last Name_2    Company  \
2           2400         True         John    Williams  Microsoft   
4           2560         True        Chloe    Anderson     Disney   
5           2160         True       Jeremy       Scott     Adidas   

             Name  
2   John Williams  
4  Chloe Anderson  
5    Jeremy Scott

## Tutorial - Flight delay data 

https://www.kaggle.com/divyansh22/flight-delay-prediction

Now, we will dive into a more complicated data analysis using Pandas.

This dataset is the flight delay prediction for the month of January. 

This data is collected from the Bureau of Transportation Statistics, Govt. of the USA. This data is open-sourced under U.S. Govt. Works. This dataset contains all the flights in the month of January 2020. There are more than 400,000 flights in the month of January itself throughout the United States. The features were manually chosen to do a primary time series analysis. There are several other features available on their website.

This data could well be used to predict the flight delay at the destination airport specifically for the month of January in upcoming years as the data is for January only.

In [101]:
df = pd.read_csv('data/Jan_2020_ontime.csv')

As always, the first thing we need to do is get an idea of what data looks like!
We can call three functions here:
- head()
- describe()
- info()

In [103]:
df

DAY_OF_MONTH  DAY_OF_WEEK OP_UNIQUE_CARRIER  OP_CARRIER_AIRLINE_ID  \
0                  1            3                EV                  20366   
1                  1            3                EV                  20366   
2                  1            3                EV                  20366   
3                  1            3                EV                  20366   
4                  1            3                EV                  20366   
...              ...          ...               ...                    ...   
607341            31            5                9E                  20363   
607342            31            5                9E                  20363   
607343            31            5                9E                  20363   
607344            31            5                9E                  20363   
607345            31            5                9E                  20363   

       OP_CARRIER TAIL_NUM  OP_CARRIER_FL_NUM  ORIGIN_AIRPORT_ID  \
0              EV   N48901               4397              13930   
1              EV   N16976               4401              15370   
2              EV   N12167               4404              11618   
3              EV   N14902               4405              10781   
4              EV   N606UX               4407              14524   
...           ...      ...                ...                ...   
607341         9E   N331CA               4812              15412   
607342         9E   N295PQ               4813              11433   
607343         9E   N294PQ               4814              11996   
607344         9E   N228PQ               4815              10397   
607345         9E   N228PQ               4815              15919   

        ORIGIN_AIRPORT_SEQ_ID ORIGIN  ...  DEST  DEP_TIME DEP_DEL15  \
0                     1393007    ORD  ...   GRB    1003.0       0.0   
1                     1537002    TUL  ...   ORD    1027.0       0.0   
2                     1161802    EWR  ...   TYS    1848.0       0.0   
3                     1078105    BTR  ...   IAH    1846.0       0.0   
4                     1452401    RIC  ...   IAH    1038.0       0.0   
...                       ...    ...  ...   ...       ...       ...   
607341                1541205    TYS  ...   DTW    1002.0       1.0   
607342                1143302    DTW  ...   JFK    1747.0       0.0   
607343                1199603    GSP  ...   LGA     554.0       0.0   
607344                1039707    ATL  ...   XNA    1714.0       0.0   
607345                1591904    XNA  ...   ATL    1841.0       0.0   

        DEP_TIME_BLK  ARR_TIME ARR_DEL15  CANCELLED  DIVERTED  DISTANCE  \
0          1000-1059    1117.0       0.0        0.0       0.0     174.0   
1          1000-1059    1216.0       0.0        0.0       0.0     585.0   
2          1800-1859    2120.0       0.0        0.0       0.0     631.0   
3          1800-1859    2004.0       0.0        0.0       0.0     253.0   
4          1000-1059    1330.0       0.0        0.0       0.0    1157.0   
...              ...       ...       ...        ...       ...       ...   
607341     0700-0759    1128.0       1.0        0.0       0.0     443.0   
607342     1700-1759    1933.0       0.0        0.0       0.0     509.0   
607343     0600-0659     752.0       0.0        0.0       0.0     610.0   
607344     1700-1759    1811.0       0.0        0.0       0.0     589.0   
607345     1800-1859    2128.0       0.0        0.0       0.0     589.0   

        Unnamed: 21  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN  
...             ...  
607341          NaN  
607342          NaN  
607343          NaN  
607344          NaN  
607345          NaN  

[607346 rows x 22 columns]

In [104]:
df.head()

DAY_OF_MONTH  DAY_OF_WEEK OP_UNIQUE_CARRIER  OP_CARRIER_AIRLINE_ID  \
0             1            3                EV                  20366   
1             1            3                EV                  20366   
2             1            3                EV                  20366   
3             1            3                EV                  20366   
4             1            3                EV                  20366   

  OP_CARRIER TAIL_NUM  OP_CARRIER_FL_NUM  ORIGIN_AIRPORT_ID  \
0         EV   N48901               4397              13930   
1         EV   N16976               4401              15370   
2         EV   N12167               4404              11618   
3         EV   N14902               4405              10781   
4         EV   N606UX               4407              14524   

   ORIGIN_AIRPORT_SEQ_ID ORIGIN  ...  DEST  DEP_TIME DEP_DEL15  DEP_TIME_BLK  \
0                1393007    ORD  ...   GRB    1003.0       0.0     1000-1059   
1                1537002    TUL  ...   ORD    1027.0       0.0     1000-1059   
2                1161802    EWR  ...   TYS    1848.0       0.0     1800-1859   
3                1078105    BTR  ...   IAH    1846.0       0.0     1800-1859   
4                1452401    RIC  ...   IAH    1038.0       0.0     1000-1059   

   ARR_TIME ARR_DEL15  CANCELLED  DIVERTED  DISTANCE  Unnamed: 21  
0    1117.0       0.0        0.0       0.0     174.0          NaN  
1    1216.0       0.0        0.0       0.0     585.0          NaN  
2    2120.0       0.0        0.0       0.0     631.0          NaN  
3    2004.0       0.0        0.0       0.0     253.0          NaN  
4    1330.0       0.0        0.0       0.0    1157.0          NaN  

[5 rows x 22 columns]

In [105]:
df.describe()

DAY_OF_MONTH    DAY_OF_WEEK  OP_CARRIER_AIRLINE_ID  OP_CARRIER_FL_NUM  \
count  607346.000000  607346.000000          607346.000000      607346.000000   
mean       16.014354       3.955735           19994.137492        2622.365261   
std         8.990719       1.910205             375.754232        1822.545302   
min         1.000000       1.000000           19393.000000           1.000000   
25%         8.000000       2.000000           19790.000000        1070.000000   
50%        16.000000       4.000000           19977.000000        2177.000000   
75%        24.000000       5.000000           20378.000000        4108.000000   
max        31.000000       7.000000           20452.000000        6860.000000   

       ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  DEST_AIRPORT_ID  \
count      607346.000000           6.073460e+05    607346.000000   
mean        12657.389167           1.265743e+06     12657.196320   
std          1524.407203           1.524405e+05      1524.279269   
min         10135.000000           1.013506e+06     10135.000000   
25%         11292.000000           1.129202e+06     11292.000000   
50%         12889.000000           1.288903e+06     12889.000000   
75%         14027.000000           1.402702e+06     14027.000000   
max         16869.000000           1.686901e+06     16869.000000   

       DEST_AIRPORT_SEQ_ID       DEP_TIME      DEP_DEL15       ARR_TIME  \
count         6.073460e+05  600682.000000  600647.000000  600271.000000   
mean          1.265724e+06    1331.512559       0.136326    1477.968924   
std           1.524277e+05     497.095168       0.343135     525.543294   
min           1.013506e+06       1.000000       0.000000       1.000000   
25%           1.129202e+06     919.000000       0.000000    1058.000000   
50%           1.288903e+06    1326.000000       0.000000    1513.000000   
75%           1.402702e+06    1739.000000       0.000000    1918.000000   
max           1.686901e+06    2400.000000       1.000000    2400.000000   

           ARR_DEL15      CANCELLED       DIVERTED       DISTANCE  Unnamed: 21  
count  599268.000000  607346.000000  607346.000000  607346.000000          0.0  
mean        0.137309       0.011407       0.001893     798.022341          NaN  
std         0.344174       0.106193       0.043473     587.282639          NaN  
min         0.000000       0.000000       0.000000      31.000000          NaN  
25%         0.000000       0.000000       0.000000     369.000000          NaN  
50%         0.000000       0.000000       0.000000     641.000000          NaN  
75%         0.000000       0.000000       0.000000    1037.000000          NaN  
max         1.000000       1.000000       1.000000    5095.000000          NaN

In [106]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 607346 entries, 0 to 607345
Data columns (total 22 columns):
DAY_OF_MONTH             607346 non-null int64
DAY_OF_WEEK              607346 non-null int64
OP_UNIQUE_CARRIER        607346 non-null object
OP_CARRIER_AIRLINE_ID    607346 non-null int64
OP_CARRIER               607346 non-null object
TAIL_NUM                 606648 non-null object
OP_CARRIER_FL_NUM        607346 non-null int64
ORIGIN_AIRPORT_ID        607346 non-null int64
ORIGIN_AIRPORT_SEQ_ID    607346 non-null int64
ORIGIN                   607346 non-null object
DEST_AIRPORT_ID          607346 non-null int64
DEST_AIRPORT_SEQ_ID      607346 non-null int64
DEST                     607346 non-null object
DEP_TIME                 600682 non-null float64
DEP_DEL15                600647 non-null float64
DEP_TIME_BLK             607346 non-null object
ARR_TIME                 600271 non-null float64
ARR_DEL15                599268 non-null float64
CANCELLED                6073

### Columns to use:
- Day_of_week: Day of Week starting from Monday
  - Ex: Monday = 1, Sunday = 7
- Op_unique_carrier: Unique Carrier Code (DL, WN, AA, 9E, OO, etc.)
- Origin: Origin Airport (JFK, ATL, SFO, LAX, etc.)
- Dest: Destination Airport (JFK, ATL, SFO, LAX, etc.)
- Dep_time: Actual Departure Time (local time: hhmm)
  - Ex: 1848 is 18:48 in local time
- Dep_del15: Departure Delay Indicator, 15 Minutes or More (1=Yes, 0=No)
- Arr_time: Actual Arrival Time (local time: hhmm)
- Arr_del15: Arrival Delay Indicator, 15 Minutes or More (1=Yes, 0=No)
- Cancelled: Cancelled Flight Indicator (1=Yes, 0=No)
- Diverted: Diverted Flight Indicator (1=Yes, 0=No)
- Distance: Distance between airports (miles)

We are going to select the columns to use using ```[]```.

In [107]:
flight = df[["DAY_OF_WEEK", "OP_UNIQUE_CARRIER", "ORIGIN", "DEST", "DEP_TIME", "DEP_DEL15", "ARR_TIME", "ARR_DEL15", "CANCELLED", "DIVERTED", "DISTANCE"]]
flight

DAY_OF_WEEK OP_UNIQUE_CARRIER ORIGIN DEST  DEP_TIME  DEP_DEL15  \
0                 3                EV    ORD  GRB    1003.0        0.0   
1                 3                EV    TUL  ORD    1027.0        0.0   
2                 3                EV    EWR  TYS    1848.0        0.0   
3                 3                EV    BTR  IAH    1846.0        0.0   
4                 3                EV    RIC  IAH    1038.0        0.0   
...             ...               ...    ...  ...       ...        ...   
607341            5                9E    TYS  DTW    1002.0        1.0   
607342            5                9E    DTW  JFK    1747.0        0.0   
607343            5                9E    GSP  LGA     554.0        0.0   
607344            5                9E    ATL  XNA    1714.0        0.0   
607345            5                9E    XNA  ATL    1841.0        0.0   

        ARR_TIME  ARR_DEL15  CANCELLED  DIVERTED  DISTANCE  
0         1117.0        0.0        0.0       0.0     174.0  
1         1216.0        0.0        0.0       0.0     585.0  
2         2120.0        0.0        0.0       0.0     631.0  
3         2004.0        0.0        0.0       0.0     253.0  
4         1330.0        0.0        0.0       0.0    1157.0  
...          ...        ...        ...       ...       ...  
607341    1128.0        1.0        0.0       0.0     443.0  
607342    1933.0        0.0        0.0       0.0     509.0  
607343     752.0        0.0        0.0       0.0     610.0  
607344    1811.0        0.0        0.0       0.0     589.0  
607345    2128.0        0.0        0.0       0.0     589.0  

[607346 rows x 11 columns]

### Join datasets
- ```flight``` dataframe has ```OP_UNIQUE_CARRIER```, but it is missing airline names.
- ```airline_names.xlsx``` has airline codes as well as airline names.

Let's join ```flight``` dataframe with ```airline_names.xlsx``` to get airline names.

In [108]:
airline = pd.read_excel('data/airline_names.xlsx')
airline

carrier                 carrier_name
0       9E            Endeavor Air Inc.
1       AA       American Airlines Inc.
2       AQ          Aloha Airlines Inc.
3       AS         Alaska Airlines Inc.
4       B6              JetBlue Airways
5       CO   Continental Air Lines Inc.
6       DL         Delta Air Lines Inc.
7       EV      ExpressJet Airlines LLC
8       F9       Frontier Airlines Inc.
9       FL  AirTran Airways Corporation
10      G4                Allegiant Air
11      HA       Hawaiian Airlines Inc.
12      HP   America West Airlines Inc.
13      MQ                    Envoy Air
14      NK             Spirit Air Lines
15      NW      Northwest Airlines Inc.
16      OH                  Comair Inc.
17      OO        SkyWest Airlines Inc.
18      RU     ExpressJet Airlines Inc.
19      TZ       ATA Airlines d/b/a ATA
20      UA        United Air Lines Inc.
21      US              US Airways Inc.
22      VX               Virgin America
23      WN       Southwest Airlines Co.
24      XE     ExpressJet Airlines Inc.
25      YV           Mesa Airlines Inc.
26      YX             Republic Airline

We are going to use ```merge``` function that we learned today and specify left column name and right column name since they are different.

In [109]:
flight_airline = pd.merge(flight, airline, left_on='OP_UNIQUE_CARRIER', right_on='carrier')
flight_airline

DAY_OF_WEEK OP_UNIQUE_CARRIER ORIGIN DEST  DEP_TIME  DEP_DEL15  \
0                 3                EV    ORD  GRB    1003.0        0.0   
1                 3                EV    TUL  ORD    1027.0        0.0   
2                 3                EV    EWR  TYS    1848.0        0.0   
3                 3                EV    BTR  IAH    1846.0        0.0   
4                 3                EV    RIC  IAH    1038.0        0.0   
...             ...               ...    ...  ...       ...        ...   
607341            5                AS    SEA  SJC    1336.0        0.0   
607342            5                AS    SJC  SEA    2001.0        1.0   
607343            5                AS    SEA  TUS    1956.0        1.0   
607344            5                AS    IAD  LAX     651.0        0.0   
607345            5                AS    JFK  PDX    1258.0        0.0   

        ARR_TIME  ARR_DEL15  CANCELLED  DIVERTED  DISTANCE carrier  \
0         1117.0        0.0        0.0       0.0     174.0      EV   
1         1216.0        0.0        0.0       0.0     585.0      EV   
2         2120.0        0.0        0.0       0.0     631.0      EV   
3         2004.0        0.0        0.0       0.0     253.0      EV   
4         1330.0        0.0        0.0       0.0    1157.0      EV   
...          ...        ...        ...       ...       ...     ...   
607341    1541.0        0.0        0.0       0.0     696.0      AS   
607342    2229.0        1.0        0.0       0.0     696.0      AS   
607343    2400.0        1.0        0.0       0.0    1216.0      AS   
607344     910.0        0.0        0.0       0.0    2288.0      AS   
607345    1605.0        0.0        0.0       0.0    2454.0      AS   

                   carrier_name  
0       ExpressJet Airlines LLC  
1       ExpressJet Airlines LLC  
2       ExpressJet Airlines LLC  
3       ExpressJet Airlines LLC  
4       ExpressJet Airlines LLC  
...                         ...  
607341     Alaska Airlines Inc.  
607342     Alaska Airlines Inc.  
607343     Alaska Airlines Inc.  
607344     Alaska Airlines Inc.  
607345     Alaska Airlines Inc.  

[607346 rows x 13 columns]

Now, ```flight_airline``` dataframe has all of the delay, cancellation info as well as airline codes and names!

## Groupby
- which airlines experience most delays? (# delays by airline)
- what day of a week experience most delays? (# delays by the day of a week)
- what origin experience most delays? (# delays by origin and destination)

### Which airlines experience most delays?

This question can be reworded as "# of delays by airline"

We can use ```groupby``` function to get the number of delays by each airline!

In [110]:
flight_airline.groupby(['carrier','carrier_name'])['DEP_DEL15'].sum().sort_values(ascending=False)

carrier  carrier_name           
WN       Southwest Airlines Co.     13381.0
AA       American Airlines Inc.     11000.0
OO       SkyWest Airlines Inc.      10836.0
DL       Delta Air Lines Inc.        8095.0
UA       United Air Lines Inc.       6165.0
OH       Comair Inc.                 4656.0
MQ       Envoy Air                   4299.0
AS       Alaska Airlines Inc.        3863.0
B6       JetBlue Airways             3735.0
YX       Republic Airline            2964.0
YV       Mesa Airlines Inc.          2802.0
9E       Endeavor Air Inc.           2580.0
NK       Spirit Air Lines            2403.0
F9       Frontier Airlines Inc.      1888.0
G4       Allegiant Air               1376.0
EV       ExpressJet Airlines LLC     1284.0
HA       Hawaiian Airlines Inc.       557.0
Name: DEP_DEL15, dtype: float64

If the threshold is 10,000 delays in January, it seems that WN, AA, OO are top three airlines that experience most delays.
But often times, passengers would care more about **arrival on time** and that's what airline industries focus more.

### 1.0 - Now Try This

Calculate sum of ARR_DEL15 using the groupby function and sort the dataset in a descending order.

**Answer**: flight_airline.groupby(['carrier'])['ARR_DEL15'].sum().sort_values(ascending=False)

If we use the same threshold of 10,000 delays, AA, OO, OH, and WN are airlines that experience most delays in regards to **arrival**.

We've found the top airlines with most delays but this result does not take into account "Total number of flights."

So, it will be more accurate if we can find the percentage of delayed flights for each airline.

```delayed flight % = (delayed flights) / (total flights)```

Let's calculate **delayed flight %**!

In [111]:
# this code returns the total number of flights
flight_airline.groupby(['carrier','carrier_name'])['ARR_DEL15'].count()

carrier  carrier_name           
9E       Endeavor Air Inc.           22955
AA       American Airlines Inc.      75472
AS       Alaska Airlines Inc.        20783
B6       JetBlue Airways             24621
DL       Delta Air Lines Inc.        79928
EV       ExpressJet Airlines LLC     10376
F9       Frontier Airlines Inc.      12312
G4       Allegiant Air                7663
HA       Hawaiian Airlines Inc.       7015
MQ       Envoy Air                   25284
NK       Spirit Air Lines            17333
OH       Comair Inc.                 23999
OO       SkyWest Airlines Inc.       69501
UA       United Air Lines Inc.       48019
WN       Southwest Airlines Co.     107708
YV       Mesa Airlines Inc.          17473
YX       Republic Airline            28826
Name: ARR_DEL15, dtype: int64

Using this formula, ```delayed flight % = (delayed flights) / (total flights)```

we will divide the number of delayed flights by total flights.

In [112]:
# this code returns the percentage of delayed flights of each airline.
flight_airline.groupby(['carrier','carrier_name'])['ARR_DEL15'].sum() / flight_airline.groupby(['carrier','carrier_name'])['ARR_DEL15'].count()

carrier  carrier_name           
9E       Endeavor Air Inc.          0.109562
AA       American Airlines Inc.     0.150360
AS       Alaska Airlines Inc.       0.211375
B6       JetBlue Airways            0.149263
DL       Delta Air Lines Inc.       0.105833
EV       ExpressJet Airlines LLC    0.154298
F9       Frontier Airlines Inc.     0.145793
G4       Allegiant Air              0.200705
HA       Hawaiian Airlines Inc.     0.098076
MQ       Envoy Air                  0.196923
NK       Spirit Air Lines           0.134656
OH       Comair Inc.                0.215259
OO       SkyWest Airlines Inc.      0.153840
UA       United Air Lines Inc.      0.129844
WN       Southwest Airlines Co.     0.095824
YV       Mesa Airlines Inc.         0.178447
YX       Republic Airline           0.118920
Name: ARR_DEL15, dtype: float64

Now, in order to see the top airlines with highest delay percentages, let's sort the result in a descending order!

In [113]:
# sort_values(ascending = False) has been added
(flight_airline.groupby(['carrier','carrier_name'])['ARR_DEL15'].sum() / flight_airline.groupby(['carrier','carrier_name'])['ARR_DEL15'].count()).sort_values(ascending = False)

carrier  carrier_name           
OH       Comair Inc.                0.215259
AS       Alaska Airlines Inc.       0.211375
G4       Allegiant Air              0.200705
MQ       Envoy Air                  0.196923
YV       Mesa Airlines Inc.         0.178447
EV       ExpressJet Airlines LLC    0.154298
OO       SkyWest Airlines Inc.      0.153840
AA       American Airlines Inc.     0.150360
B6       JetBlue Airways            0.149263
F9       Frontier Airlines Inc.     0.145793
NK       Spirit Air Lines           0.134656
UA       United Air Lines Inc.      0.129844
YX       Republic Airline           0.118920
9E       Endeavor Air Inc.          0.109562
DL       Delta Air Lines Inc.       0.105833
HA       Hawaiian Airlines Inc.     0.098076
WN       Southwest Airlines Co.     0.095824
Name: ARR_DEL15, dtype: float64

Tada, we've come up with a completely different result!

OH, AS, G4 airlines have highest delay percentages while 9E, DL, HA, WN have reletively low delay percentages.

#### agg()

The code above is correct but it's very lengthy because we calculate ```sum()``` and ```count()``` separately.

We will learn a shortcut to calculate multiple aggregate functions using ```agg()``` argument.

Everything remains same, except that we replace ```sum()``` or ```count()``` with ```agg(['sum','count'])```.

In [114]:
new_df = flight_airline.groupby(['carrier','carrier_name'])['ARR_DEL15'].agg(['sum','count'])
new_df

sum   count
carrier carrier_name                            
9E      Endeavor Air Inc.         2515.0   22955
AA      American Airlines Inc.   11348.0   75472
AS      Alaska Airlines Inc.      4393.0   20783
B6      JetBlue Airways           3675.0   24621
DL      Delta Air Lines Inc.      8459.0   79928
EV      ExpressJet Airlines LLC   1601.0   10376
F9      Frontier Airlines Inc.    1795.0   12312
G4      Allegiant Air             1538.0    7663
HA      Hawaiian Airlines Inc.     688.0    7015
MQ      Envoy Air                 4979.0   25284
NK      Spirit Air Lines          2334.0   17333
OH      Comair Inc.               5166.0   23999
OO      SkyWest Airlines Inc.    10692.0   69501
UA      United Air Lines Inc.     6235.0   48019
WN      Southwest Airlines Co.   10321.0  107708
YV      Mesa Airlines Inc.        3118.0   17473
YX      Republic Airline          3428.0   28826

Now, ```new_df``` has columns: ```carrier```, ```carrier_name```, ```sum```, and ```count```.

We will create another column called ```delay %``` using sum and count columns!

In [115]:
new_df['delay %'] = new_df['sum'] / new_df['count']

In [116]:
new_df

sum   count   delay %
carrier carrier_name                                      
9E      Endeavor Air Inc.         2515.0   22955  0.109562
AA      American Airlines Inc.   11348.0   75472  0.150360
AS      Alaska Airlines Inc.      4393.0   20783  0.211375
B6      JetBlue Airways           3675.0   24621  0.149263
DL      Delta Air Lines Inc.      8459.0   79928  0.105833
EV      ExpressJet Airlines LLC   1601.0   10376  0.154298
F9      Frontier Airlines Inc.    1795.0   12312  0.145793
G4      Allegiant Air             1538.0    7663  0.200705
HA      Hawaiian Airlines Inc.     688.0    7015  0.098076
MQ      Envoy Air                 4979.0   25284  0.196923
NK      Spirit Air Lines          2334.0   17333  0.134656
OH      Comair Inc.               5166.0   23999  0.215259
OO      SkyWest Airlines Inc.    10692.0   69501  0.153840
UA      United Air Lines Inc.     6235.0   48019  0.129844
WN      Southwest Airlines Co.   10321.0  107708  0.095824
YV      Mesa Airlines Inc.        3118.0   17473  0.178447
YX      Republic Airline          3428.0   28826  0.118920

Tada! We've successfully added a ```delay %``` column! Let's sort the dataframe in a descending order to find out airlines with highest delay %!

In [117]:
new_df.sort_values(by=['delay %'], ascending=False)

sum   count   delay %
carrier carrier_name                                      
OH      Comair Inc.               5166.0   23999  0.215259
AS      Alaska Airlines Inc.      4393.0   20783  0.211375
G4      Allegiant Air             1538.0    7663  0.200705
MQ      Envoy Air                 4979.0   25284  0.196923
YV      Mesa Airlines Inc.        3118.0   17473  0.178447
EV      ExpressJet Airlines LLC   1601.0   10376  0.154298
OO      SkyWest Airlines Inc.    10692.0   69501  0.153840
AA      American Airlines Inc.   11348.0   75472  0.150360
B6      JetBlue Airways           3675.0   24621  0.149263
F9      Frontier Airlines Inc.    1795.0   12312  0.145793
NK      Spirit Air Lines          2334.0   17333  0.134656
UA      United Air Lines Inc.     6235.0   48019  0.129844
YX      Republic Airline          3428.0   28826  0.118920
9E      Endeavor Air Inc.         2515.0   22955  0.109562
DL      Delta Air Lines Inc.      8459.0   79928  0.105833
HA      Hawaiian Airlines Inc.     688.0    7015  0.098076
WN      Southwest Airlines Co.   10321.0  107708  0.095824

### What day of a week experience most delays? 

Let's find out what day of a week experience most delays!

In the data world, this question can be answered if we can find ```# delays by the day of a week```.

To answer this question, we can use groupby function, but should we use ```sum```? ```count```? or ```mean```?

- ```sum```: The first thing that came to my mind is that we can use ```sum```. But similar to the question above, each day of a week has a different number of flights each day. (Ex: there would be more flights scheduled on weekends.) ```Sum``` is not the most appropriate function to use here.


- ```count```: We can count the number of flights and divide ```sum(delayed flights)``` by ```count``` (similar to what we just did)


- ```mean```: What if we use ```mean```? We can get the average delay percentage of each day and this is what we are looking for!



In [118]:
flight_airline.groupby(['DAY_OF_WEEK'])['ARR_DEL15'].mean()

DAY_OF_WEEK
1    0.133963
2    0.098135
3    0.097656
4    0.137520
5    0.161769
6    0.211572
7    0.135598
Name: ARR_DEL15, dtype: float64

To confirm that ```mean``` returns same value as ```sum / count```, let's check ```sum / count``` real quick.

In [119]:
flight_airline.groupby(['DAY_OF_WEEK'])['ARR_DEL15'].sum() / flight_airline.groupby(['DAY_OF_WEEK'])['ARR_DEL15'].count()

DAY_OF_WEEK
1    0.133963
2    0.098135
3    0.097656
4    0.137520
5    0.161769
6    0.211572
7    0.135598
Name: ARR_DEL15, dtype: float64

Yay! It returns same answer as above!

Saturday has the highest delay percentage while Wednesday has the lowest delay percentage!

### What origins experience most delays? 

Let's find out what origins experience most delays?

Knowing this might help when we book a flight later! :)

Since we are dealing with **Origin** airport this time, it's appropriate to use ```DEP_DEL15```

In [120]:
flight_airline.groupby(['ORIGIN'])['DEP_DEL15'].mean()

ORIGIN
ABE    0.178273
ABI    0.190184
ABQ    0.099070
ABR    0.250000
ABY    0.059524
         ...   
WRG    0.189655
XNA    0.119005
XWA    0.238806
YAK    0.211538
YUM    0.139785
Name: DEP_DEL15, Length: 351, dtype: float64

In [121]:
flight_airline.groupby(['ORIGIN'])['DEP_DEL15'].mean().sort_values(ascending=False)

ORIGIN
ADK    0.333333
OTH    0.333333
OGD    0.333333
PPG    0.333333
ASE    0.303813
         ...   
CPR    0.038095
CDC    0.037037
UIN    0.000000
COD    0.000000
OWB    0.000000
Name: DEP_DEL15, Length: 351, dtype: float64

This resulting Series show Origin airports and the corresponding delay %.

Since we don't know all of the airport names from the codes, we can join this Series to the dataset that has airport codes and airport names.

## Filtering based on condition

- Which airports have delay percentages of 30% or more?

### Which airports have delay percentages of 30% or more?

```airline_delay_causes.csv``` has lots of useful information, but we will extract airport codes and airport names only.

In [122]:
airport_codes = pd.read_csv('data/airline_delay_causes.csv')

In [123]:
airport_codes.head()

year   month carrier          carrier_name airport  \
0  2004       1      DL  Delta Air Lines Inc.     PBI   
1  2004       1      DL  Delta Air Lines Inc.     PDX   
2  2004       1      DL  Delta Air Lines Inc.     PHL   
3  2004       1      DL  Delta Air Lines Inc.     PHX   
4  2004       1      DL  Delta Air Lines Inc.     PIT   

                                        airport_name  arr_flights  arr_del15  \
0  West Palm Beach/Palm Beach, FL: Palm Beach Int...        650.0      126.0   
1               Portland, OR: Portland International        314.0       61.0   
2       Philadelphia, PA: Philadelphia International        513.0       97.0   
3      Phoenix, AZ: Phoenix Sky Harbor International        334.0       78.0   
4           Pittsburgh, PA: Pittsburgh International        217.0       47.0   

   carrier_ct   weather_ct  ...  late_aircraft_ct  arr_cancelled  \
0       21.06         6.44  ...             45.92            4.0   
1       14.09         2.61  ...             10.05           30.0   
2       27.60         0.42  ...             17.12           15.0   
3       20.14         2.02  ...             16.45            3.0   
4        8.08         0.44  ...             16.59            4.0   

   arr_diverted   arr_delay   carrier_delay  weather_delay  nas_delay  \
0           0.0      5425.0           881.0          397.0     2016.0   
1           3.0      2801.0           478.0          239.0     1365.0   
2           0.0      4261.0          1150.0           16.0     2286.0   
3           1.0      3400.0          1159.0          166.0     1295.0   
4           1.0      1737.0           350.0           28.0      522.0   

   security_delay  late_aircraft_delay  Unnamed: 21  
0            15.0               2116.0          NaN  
1             0.0                719.0          NaN  
2             0.0                809.0          NaN  
3             0.0                780.0          NaN  
4             0.0                837.0          NaN  

[5 rows x 22 columns]

### 2.0 - Now Try This

Using ```drop_duplicates()``` function, find a unique set of airport codes and airport names and save the dataframe as ```airport_names```.

In [124]:
airport_names = airport_codes[["airport", "airport_name"]].drop_duplicates()
airport_names

airport                                       airport_name
0          PBI  West Palm Beach/Palm Beach, FL: Palm Beach Int...
1          PDX               Portland, OR: Portland International
2          PHL       Philadelphia, PA: Philadelphia International
3          PHX      Phoenix, AZ: Phoenix Sky Harbor International
4          PIT           Pittsburgh, PA: Pittsburgh International
...        ...                                                ...
147092     DUT                     Unalaska, AK: Unalaska Airport
150875     HKY                      Hickory, NC: Hickory Regional
157316     MTH  Marathon, FL: The Florida Keys Marathon Intern...
160027     GLH               Greenville, MS: Greenville Mid Delta
162721     MKC       Kansas City, MO: Charles B. Wheeler Downtown

[409 rows x 2 columns]

The Series below is what we have found earlier -- delay % by each origin airport.

In [125]:
series = flight_airline.groupby(['ORIGIN'])['DEP_DEL15'].mean().sort_values(ascending=False)

Let's convert the result of groupby aggregation to ```DataFrame``` using ```to_frame()``` function.

In [126]:
origin_delays = series.to_frame()

In [127]:
origin_delays

DEP_DEL15
ORIGIN           
ADK      0.333333
OTH      0.333333
OGD      0.333333
PPG      0.333333
ASE      0.303813
...           ...
CPR      0.038095
CDC      0.037037
UIN      0.000000
COD      0.000000
OWB      0.000000

[351 rows x 1 columns]

Now, let's take a closer look at the question and break it down to smaller subproblems!

**Which airports have delay percentages of 30% or more?**
1. Find airport codes with a delay percentage greater than 30%.
2. Join the resulting dataframe to ```airport_names``` to get airport names.

In [128]:
airports_most_delays = origin_delays[origin_delays['DEP_DEL15'] > 0.3]
airports_most_delays

DEP_DEL15
ORIGIN           
ADK      0.333333
OTH      0.333333
OGD      0.333333
PPG      0.333333
ASE      0.303813

In [129]:
pd.merge(airports_most_delays, airport_names, left_on="ORIGIN", right_on = "airport")

DEP_DEL15 airport                                       airport_name
0   0.333333     ADK                              Adak Island, AK: Adak
1   0.333333     OTH  North Bend/Coos Bay, OR: Southwest Oregon Regi...
2   0.333333     OGD                          Ogden, UT: Ogden-Hinckley
3   0.333333     PPG             Pago Pago, TT: Pago Pago International
4   0.303813     ASE         Aspen, CO: Aspen Pitkin County Sardy Field

We've found that ADK, OTH, OGD, PPG, and ASE have highest delay percentages in the US!

### Pop-up Question

What if we want to find the delay percentage of a specific airport?

Then we can select the origin airport code by using ```df[df[column_name]=="ABC"]```

Let's check out the delay % of San Francisco airport.

In [130]:
origin_delays[origin_delays['ORIGIN'] == 'SFO']

KeyError: 'ORIGIN'

The code above did not work! Why? It's because ```ORIGIN``` is not a column, but an index.

In the case where our selected column is an index, we can use ```index``` to select the index.

In [131]:
origin_delays[origin_delays.index == 'SFO']

DEP_DEL15
ORIGIN           
SFO      0.171547

## Resources
- https://www.geeksforgeeks.org/python-pandas-series/
- https://www.kaggle.com/divyansh22/flight-delay-prediction
- https://www.kaggle.com/anshuls235/airline-delay-causes
- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html
- https://www.codespot.org/sql-join/


## Homework
1. Find out five largest airlines by the number of flights in January
2. Calculate cancellation percentages of the top 5 airlines
3. Find out the percentage of flights that managed to arrive within 15 minutes even though the flight departed more than 15 minutes past scheduled departure time (Hint: you may use ```DEP_DEL15 == 1``` and ```ARR_DEL15 == 0```)
4. Convert categorical to numerical to do data visualizations of each airline
